### ISSUE TITLE
cuts words instead of silence

### ISSUE THREAD

**Issue by Sankyuubigan** [WyattBlue/auto-editor#582]
I Exporting to:
 - [x] Media File (Video or Audio)

Hi. the default automatic silence removal for some reason sometimes cuts words (word endings), sometimes half of the word is cut. i couldn't find a subcommand for the video editor to cut less silence, maybe leave 0.2-0.5 seconds more after each cut? my video has a static picture, I only need to crop the silence on the sound.

 - recommendedTimebase: 25/1
 - video:
   - track 0:
     - codec: h264
     - fps: 25
     - resolution: 1728x1080
     - aspect ratio: 8:5
     - pixel aspect ratio: 1
     - duration: 4078.04
     - pix fmt: yuv420p
     - color range: 1 (tv)
     - color space: 1 (bt709)
     - color primaries: 1 (bt709)
     - color transfer: 1 (bt709)
     - timebase: 1/12800
     - bitrate: 311366
     - lang: und
 - audio:
   - track 0:
     - codec: aac
     - layout: stereo
     - samplerate: 44100
     - duration: 4078.0974149659864
     - bitrate: 128000
     - lang: eng
 - container:
   - duration: 4078.097415
   - bitrate: 445713

**Comment by Sankyuubigan** [comment]
I've already had several videos lose meaning because of this, stumps of different words joined together. you need to fine tune the silence cutoff.

**Comment by utopiatopia** [comment]
The `--margin` flag solves this. Usage is in the [Cutting section of README.md](https://github.com/WyattBlue/auto-editor#cutting).

As for adjusting the margin time, I believe it's best set where it is. Most of the videos I've used are fine. However, I definitely would support making the documentation of the flags more clear (especially `--edit`)

**Comment by Sankyuubigan** [comment]
nice thx. another question - what methods are used to determine silence? a simple method only from ffmpeg? it seems to be simple, it defines silence as below the threshold value in decibels. it would explain why it cuts some words.

**Comment by biggudawgg** [comment]
ive had the same issue and the margin didnot fix it

**Comment by utopiatopia** [comment]
> what methods are used to determine silence?

AFAIK (I’m not the maintainer), there seems to be 3 ways (specified in [auto-editor/analyze.py](https://github.com/WyattBlue/auto-editor/blob/master/auto_editor/analyze.py))

As you said, determining audio levels. On a side note, it doesn’t operate on dB directly. It operates on the actual amplitude. There’s a conversion formula somewhere that does that.

Another is how much change (motion, basically) there is in the video. If a lot of pixels change (probably due to motion), it decides that the video is not “silent”.

The last way I’m aware of is subtitles. Whenever subtitles are present, it doesn’t edit that part (or treats it as not “silent”).

For how to use them, it’s also in the [README.md](https://github.com/WyattBlue/auto-editor#methods-for-making-automatic-cuts), although information about it is scarce. I’ll give it more attention when I have more time.

**Comment by utopiatopia** [comment]
> ive had the same issue and the margin didnot fix it

Try increasing the audio threshold. (e.g. `--edit audio:threshold=0.1`)

If this doesn’t help, could you provide a short sample of the problematic audio?

**Comment by Sankyuubigan** [comment]
> If this doesn’t help, could you provide a short sample of the problematic audio?

I suspect that the problem may be related to poor quality of speech recording. i.e. not very good microphone or noise. why auto-editor doesn't suppress noise before cutting silence? or is noise removal always related to a specific language? in any case, it turns out that using auto-editor at the moment with mediocre quality audio is impossible. it cuts whole phrases in my case. it loses context in my video, unfortunately.

**Comment by Sankyuubigan** [comment]
> ive had the same issue and the margin didnot fix it

I managed to slightly improve the result with this code, which I used to partially reduce the noise BEFORE using auto-editor. the result still cuts words, but now less, the context is preserved a little better.

`ffmpeg -i %file_name% -af "arnndn=m='rnnoise-models/beguiling-drafter-2018-08-30/bd.rnnn'" %file_name:.=_denoised2.%`

this noise reduction repository is already 6 years out of date. 
https://github.com/GregorR/rnnoise-models
Are there any more effective open access methods for noise removal at the moment ?

**Comment by Sankyuubigan** [comment]
> Try decreasing the audio threshold. (e.g. `--edit audio:threshold=0.02`)

What settings can I change in the auto-editor and what settings would be best for me to change in order to “soften the boundaries” of silence removal a little bit? Maybe I should expand all possible settings a little bit? I am ready to compromise in order to keep some silence in my videos. the main thing is that the words are not cut. does the auto-editor not have any “modes”, for example some gentle mode? Are there any other parameters besides margin and threshold?

**Comment by utopiatopia** [comment]
> how can I make the algorithm not look at the motion in the video at all?

It doesn’t. It only looks at the audio, unless you specify otherwise (i.e. don’t use `--edit video`)

> I suspect [auto-editor] doesn't add subtitles

Correct. There’s an issue in this repo about it though. If you want subtitles, use something like whispercpp before running auto-editor

> I suspect that the problem may be related to poor quality of speech recording.

That’s my best guess. Although I have a new theory on what’s happening. It’s possible your audio is too quiet, or the speech is too inconsistent in terms of volume. Try normalizing your audio, reducing noise (e.g. rnnoise), and see what you get.

> What settings can I change in the auto-editor and what settings would be best for me to change 

I found [this](https://github.com/WyattBlue/auto-editor/blob/master/auto_editor/help.py). `--edit-based-on` might help, if you want to do some manual editing.

> does the auto-editor not have any “modes”

Not that I’m aware of.

**Comment by Sankyuubigan** [comment]
> That’s my best guess. Although I have a new theory on what’s happening. It’s possible your audio is too quiet, or the speech is too inconsistent in terms of volume. Try normalizing your audio, reducing noise (e.g. rnnoise), and see what you get.

that's right, I have too quiet sound and speech is too inconsistent in volume, that's exactly what it is. what do you mean by normalization ? using dynamic sound normalization I do it after cutting silence, and I can't do it before, because then auto-editor cuts silence very badly (silence becomes very loud). or do you mean that I need to set a single volume standard first before submitting to auto-editor ? but in that case I will still have inconsistent volume.

> I found [this](https://github.com/WyattBlue/auto-editor/blob/master/auto_editor/help.py). `--edit-based-on` might help, if you want to do some manual editing.

I didn't know that the full documentation is only in this help file, thanks. i will refer to this file as the documentation for this repository, i didn't realize it myself.

**Comment by utopiatopia** [comment]
My hypothesis seems right. I’d further decrease the audio threshold until it doesn’t chop words off, then increase margin until the ends aren’t cut off. If the podcast is actually intelligible (kinda surprised it is), this should work.


If you’re not happy with the result, lowering the noise floor, running dynamic audio normalization, and removing residual noise afterwards is your better option.

**Comment by Sankyuubigan** [comment]
> Also, I’m sorry for leading you a little astray! I’m a little scatterbrained...

I guess you mean that by default the auto-editor doesn't actually analyze the motion in the video?
